Show proper numebr of line and filename on c/c++ compilatio errors

Example of misleading errors:

    $ gwm castle-on-the-grid.cpp
    Running: g++ --std=c++17 -fsanitize=address -Wall -Wextra -pthread -g -O0 /tmp/snip.LffS40-output.cpp
    /tmp/snip.LffS40-output.cpp: In function ‘int getX(const Vertex&)’:
    /tmp/snip.LffS40-output.cpp:26:15: error: ‘asastoi’ is not a member of ‘std’
       26 |   return std::asastoi(vertex.substr(0,pos));
          |               ^~~~~~~
    /tmp/snip.LffS40-output.cpp: In member function ‘int snip::BreadthFirstTree<GraphAdjacencyList, VertexType, Tree>::getDistance(const VertexType&)’:
    /tmp/snip.LffS40-output.cpp:137:12: error: ‘couadsnt’ was not declared in this scope; did you mean ‘count’?
      137 |     return couadsnt;
          |            ^~~~~~~~
          |            count
    /tmp/snip.LffS40-output.cpp: In function ‘int minimumMoves(std::vector<std::__cxx11::basic_string<char> >, int, int, int, int)’:
    /tmp/snip.LffS40-output.cpp:152:33: error: ‘makeBreadthdrstTree’ is not a member of ‘snip’; did you mean ‘makeBreadthFirstTree’?
      152 |   auto breadthFirstTree = snip::makeBreadthdrstTree(createGraph(grid, startX, startY, goalX, goalY), createVertex(startX, startY));
          |                                 ^~~~~~~~~~~~~~~~~~~
          |                                 makeBreadthFirstTree


As `snip` creates a new file, the error doesn't show the proper file name (eg: it is converted to _/tmp/snip.LffS40-output.cpp_ instead of _castle-on-the-grid.cpp_) and the number of lines also are manged.

--- a/snip.sh
+++ b/snip.sh
@@ -119,11 +119,11 @@ __snip__replace_snips() {
   local root_filename="${filename%.*}"
   local extension="${filename#"$root_filename"}"
 
-  local prefix_tmp
-  prefix_tmp=$(command -p mktemp -t snip.XXXXXX) || return 1
-  # remove old snip tmp files and the one just created as we only need the file name.
+  # remove old snip tmp files
   # using --dry-run is unsafe. not doing so, we also test access to tmp filesystem. see man mktemp
-  rm "${prefix_tmp%/*}"/snip.*
+  local tmp_file
+  tmp_file=$(command -p mktemp -t snip.XXXXXX) || return 1
+  rm "${tmp_file%/*}"/snip.*
 
   local cache_dir=~/.cache/snip
   mkdir -p "$cache_dir"/
@@ -148,12 +148,12 @@ __snip__replace_snips() {
     # replace snips recursively
     local recursive_snippet_file
     recursive_snippet_file=$(__snip__replace_snips "$snippet_file" $force) || return 1
-    local new_file=$prefix_tmp-$i--$filename
+    local new_file=$tmp_file-$i--$filename
     sed -r "\@$snippet@r"<( cat "$recursive_snippet_file" ) "$source_file" > "$new_file" || return 1
     source_file="$new_file"
   done
 
-  local output_file=${prefix_tmp}-output${extension}
+  local output_file=${tmp_file}-output${extension}
   __snip__remove_snip_lines "$source_file" > "$output_file"
   echo "$output_file"
 }

--- a/snip.sh
+++ b/snip.sh
@@ -102,7 +102,8 @@ __snip__replace_snips() {
   local force
   force=$2
 
-  mapfile -t snippets < <(grep -Po '(^|(?<=[^[:alnum:]]))(?<=snip\(")[^"]+' "$source_file")
+  # get snippet urls/paths with line numbers as a prefix
+  mapfile -t snippets < <(grep -nPo '(^|(?<=[^[:alnum:]]))(?<=snip\(")[^"]+' "$source_file")
   local n_snippets=${#snippets[@]}
   if (( n_snippets == 0 )); then
     echo $source_file
@@ -130,7 +131,7 @@ __snip__replace_snips() {
 
   local i
   for ((i=0; i < n_snippets; ++i)); do
-    local snippet="${snippets[$i]}"
+    local snippet="${snippets[$i]/*:/}"
 
     local snippet_file
     if __snip__is_local_snippet "$snippet"; then

--- a/snip.sh
+++ b/snip.sh
@@ -126,6 +126,13 @@ __snip__replace_snips() {
   tmp_file=$(command -p mktemp -t snip.XXXXXX) || return 1
   rm "${tmp_file%/*}"/snip.*
 
+  # show proper number of line and filename on c/c++ compilation errors
+  cp "$source_file" "$tmp_file"
+  if __snip__is_c_or_cpp_file "$source_file"; then
+    __snip__process_c_or_cpp_file "$source_file" "$tmp_file" "${snippets[@]}" || return 1
+  fi
+  source_file=$tmp_file
+
   local cache_dir=~/.cache/snip
   mkdir -p "$cache_dir"/
 
@@ -170,6 +177,36 @@ __snip__is_text_file() {
   local filename
   filename=${1:?Missing filename by param}
   __snip__is_regular_file "$filename" && [[ $(file -i -- "$filename" 2> /dev/null) =~ text/ ]]
+}
+
+
+__snip__is_c_or_cpp_file() {
+  local filename
+  filename=${1:?Missing filename by param}
+  [[ $(file -i -- "$filename" 2> /dev/null) =~ text/x-c[\+\;] ]]
+}
+
+
+__snip__process_c_or_cpp_file() {
+  local source_file
+  source_file=${1:?Missing source file as param}
+  shift
+  local tmp_file
+  tmp_file=${1:?Missing tmp file as param}
+  shift
+  : ${1:?Missing snippets as params}
+
+  sed -i.bak \
+    "1i #line 1 \"$source_file\"" \
+    "$tmp_file" || return 1
+
+  local snippet
+  for snippet in "$@"; do
+    sed -i.bak \
+      -e "${snippet/:*/}i #line 1 \"${snippet/*:/}\"" \
+      -e "${snippet/:*/}a #line $((${snippet/:*/} + 1)) \"$source_file\"" \
+      "$tmp_file" || return 1
+  done
 }
 
 

